File: packages/api/src/lua/joblock.lua
-- keys: 1
--
-- Acquire a job lock, or report who holds it.
--
-- Scheduled jobs must run exactly once per tick across every instance
-- (NFR-8). Several API or job processes all wake on the same cron
-- schedule, and without a lock every one of them would apply a drift
-- tick - so prices would move once per instance rather than once per
-- tick, and the drift bounds would be violated by exactly the number of
-- machines deployed.
--
-- SET NX PX is a single atomic operation, so only one caller can win.
-- The PX expiry is what makes it safe: a process that dies holding the
-- lock does not wedge the job forever.
--
-- The token matters on release. Without it, a process whose lock has
-- already expired - because it ran long - would delete a lock a *second*
-- process legitimately acquired, and then both would run. Release only
-- deletes the lock if the token still matches.

local key   = KEYS[1]
local token = ARGV[1]
local ttl   = tonumber(ARGV[2])

if redis.call('SET', key, token, 'NX', 'PX', ttl) then
  return 1
end
return 0

File: packages/api/src/services/jobLock.js
import { randomUUID } from 'node:crypto';
import { getRedis } from '../redis/client.js';
import { log } from '../log.js';

/**
 * Run a function only if this process wins the lock for it.
 *
 * NFR-8: scheduled jobs run exactly once per tick across all instances.
 * Without this, every deployed instance would apply its own drift tick
 * and prices would move N times per tick for N machines.
 */
export async function withJobLock(name, ttlMs, fn) {
  const key = `lock:job:${name}`;
  const token = randomUUID();
  const redis = getRedis();

  const acquired = await redis.joblock(key, token, ttlMs);
  if (acquired !== 1) {
    log.debug('job lock held elsewhere, skipping', { job: name });
    return { ran: false };
  }

  try {
    const result = await fn();
    return { ran: true, result };
  } finally {
    await redis.jobunlock(key, token);
  }
}
